File: apps/android/app/src/main/java/com/litter/android/state/VoiceTranscriptionPolicy.kt
package com.litter.android.state

import uniffi.codex_mobile_client.AppServerTransportState

/**
 * Voice transcription opens an Alleycat-side JSONL RPC. Treat only a fully
 * connected transport as ready for immediate transcription.
 */
internal fun shouldReconnectBeforeVoiceTranscription(state: AppServerTransportState): Boolean =
    state != AppServerTransportState.CONNECTED

/**
 * Retry voice transcription only for transport-level connectivity failures.
 * ASR credential, model, and server errors must surface unchanged.
 */
internal fun shouldRetryVoiceTranscriptionAfterReconnect(error: Throwable): Boolean {
    val message = error.message?.lowercase().orEmpty()
    if ("transport error" !in message) return false
    return "disconnected" in message ||
        "not connected" in message ||
        "connection failed" in message ||
        "timed out" in message
}

File: apps/ios/Sources/Litter/Models/VoiceTranscriptionPolicy.swift
import Foundation

/// Voice transcription opens a paired-host RPC. Only a connected transport is
/// ready for immediate transcription; other states should reconnect first.
func shouldReconnectBeforeVoiceTranscription(_ state: AppServerTransportState) -> Bool {
    state != .connected
}

/// Retry only transport-level failures after reconnecting. Auth, model, and
/// ASR service errors are surfaced unchanged so they can be fixed directly.
func shouldRetryVoiceTranscriptionAfterReconnect(_ error: Error) -> Bool {
    let message = error.localizedDescription.lowercased()
    guard message.contains("transport error") else { return false }
    return message.contains("disconnected")
        || message.contains("not connected")
        || message.contains("connection failed")
        || message.contains("timed out")
}
